### ISSUE TITLE
[Important] Critical racing condition on Stub Server while registering the calls

### ISSUE THREAD

**Issue by michelzanini** [mkotsur/restito#74]
Hi,

I have been using this library for a few years and I really enjoyed it. 
However, there has been cases where the tests would randomly fail when doing assertions if a call was made.

I have never had time to properly investigate so today I decided to give some time to investigate deeply.
I was able to find the issue and it's on the StubServer class, where the requests are received on method `stubsToHandler`.

In that class, there's a part that writes the response of a call to the user if the call is found on a stub with `stub.apply(response)`. And then after that it does:

```
if (isRegisterCalls()) {
    calls.add(call);
}
```

The problem here is that the call is added to the list after the response has been applied. In some VERY RARE cases, if the computer has a fast processor and you run thousands of tests, the response is written fast, the test resumes and checks if the call happened, all while the thread that finished `stub.apply(response)` has not yet reached `calls.add(call)`.

So in a simple test like this one:

```
whenHttp(server).match(get("/")).then(ok());
expect().statusCode(200).get("/");
assertEquals(1, server.getStubs().size());
```

It is possible that the `assertEquals` will happen before the `calls.add(call)`. For that to happen is very rare, specially on a small request like this one, but I was able to reproduce the problem with this test case:

```
@Test
public void reproduceRacingCondition() {
    whenHttp(server)
        .match(startsWithUri("/test-large"))
        .then(resourceContent("large-content.json"));

    for (int i = 0; i < 20000; i++) {
        expect()
            .header("Content-Type", is("application/json"))
            .header("Content-Length", is(not(nullValue())))
            .when().get("/test-large");

        assertEquals(1, server.getCalls().size());

        server.clearCalls();
    }
}
```

When you run this `20000` times, at least one time the problem is reproduced. If it does not reproduce, the tests passes in around 2 minutes. More often than not it fails. Almost every time.

To fix this issue the only change is to move the block:

```
if (isRegisterCalls()) {
    calls.add(call);
}
```

To happen before processing the stubs.

I have done this on a fork of this repository and the reproducing test always passes now.

I consider this to be a CRITICAL bug, and it has a simple fix which I can PR to your repository.

Here comes one problem, however, the master branch has changed to Java 10, not sure why. I need this on Java 8 as many people would. So I took a branch from the `0.9.3` commit.

What I ask is if you could create a branch `0.9.x` from the commit `444896fc08aa2ee11199f50bcd24789fdb42a787` : `0.9.3 release` (the same I did). I can create a PR to that branch and you can release `0.9.4` with this fix from there. That's if you do plan to change the Java version on this library, which should then, in the future, at least generate a versions such as `0.10.0` or `1.0.0`. 

If you intend to continue with Java 8 and the master branch I can try to help fixing the master branch to Java 8 etc... 
But would prefer that you do a "hot fix" branch as its easier for me.

This should only take a few minutes of your time.

Create the `0.9.x` branch, let me know, I submit the PR, you merge and release `0.9.4`. This would help me a lot.

Thanks.

**Comment by michelzanini** [comment]
This is the commit I have ready to go that fixes the issue https://github.com/mkotsur/restito/commit/c4fa84112f741980ea2f51201d23dbba75e920b4.

But I can't PR this to master as I checked out the `0.9.3` commit.

**Comment by mkotsur** [comment]
Hi Michel. Thanks for your detailed investigation. I’m on holiday now without laptop and decent internet access. As soon as I’m back, I’ll review and merge your PR.

Cheers,
Mike

**Comment by mkotsur** [comment]
Hey Michel, I'm back. And the branch https://github.com/mkotsur/restito/tree/0.9.x is there. Please submit your PR.

**Comment by michelzanini** [comment]
Here https://github.com/mkotsur/restito/pull/75

**Comment by mkotsur** [comment]
Fixed in 0.9.4

**Comment by mkotsur** [comment]
This is a good observation, @chadlwilson. I think, you are right. A `delay` action can make the issue described by @michelzanini appear more often. It seems to me, after looking at the code again, that the real issue here is that the event of Grizzly http server flushes the response may very well happen before the end of the `service()` method, which has been modified by @michelzanini 's fix.

Presumably, the response is fhushed as soon as its body is written (e.g. `resourceContent` action). In that case, the fix also works in case of delayed actions, because: 
a) registering the call happens on the same thread as the actions;
b) thanks to the fix the actions (delay and/or writing the body) always happen *after* the registering;

That's my understanding at the moment. If the problem persists, happy to accept a PR or [help to] investigate further.
